### ISSUE TITLE
Adam optimizer behavior between W10/Dell and Ubuntu/DGX

### ISSUE THREAD

**Issue by guatavita** [tensorflow/tensorflow#38605]
Hello,

I am working on semantic segmentation of 2D/3D images for medical applications and I have a Dell workstation with a Quadro GV100 32Go for development and DGX servers (workstation and server) with 4 and 8 Tesla V100 DGXS (32Go) for model deployment, training, parameters optimization, ...

The Dell and the DGX's use Tensorflow 1.15.x with the only differences being that the Dell run Windows 10 and the DGX use GNU/Linux 4.15 and the models are trained within Docker images of Ubuntu 18.0x provided by [NVIDIA](https://docs.nvidia.com/deeplearning/frameworks/tensorflow-release-notes/index.html). I think the CUDA library are different between the workstation and NVIDIA servers but both are using CUDA 10.

I am currently working on the Tensorflow implementation of the DeepLabV3+ for the segmentation of background + 12 classes. And my data consist of CT scan slices of around 55,000 images+labels for the training and 13,000 images+labels for the validation set. 

The issue that I have is that for the exact same codes, data and model architecture, the Adam optimizer behavior is completely different between the Dell and DGX. The difference is that on the DGX the validation loss increases after 3-6 epochs (depending on the learning rate) and reach a plateau with give us a prediction of an empty label for each class. The training accuracy/loss is however exactly the same between the Dell and DGX. On the Dell, the validation accuracy/loss increase/decrease smoothly and is stable, never get close to zero/reach a high plateau. Again, same exact code, data, data_generator, no data augmentation... no weird value in the images/label on the Dell workstation nor on the RAID hard drive on any of the DGX. Behavior is the same for the DGX workstation and DGX server 1.

[Tensorboard screenshot of training/validation](https://drive.google.com/file/d/1hIv4r70QHH7Ukg5SJknfkzrsS2dq6ZOq/view?usp=sharing)

I have tried to simplify the model by working on other data with only 1 label and small FOV around this label. The behavior is exactly the same and as soon as we use a LR > 0.0001 the validation accuracy drops to zero on the DGX. If we specify the LR < 0.00005 the validation accuracy increase but is still VERY unstable. 

In the end, I switched to SGD optimizer and there is no difference in term of training and validation accuracy/loss between the Dell and the DGX workstation anymore, regardless of fixed LR value or LR scheduler pattern. However, I would like to use the Adam as well and I cannot afford to no use the DGX because of that. I did not try another model such a 2D UNet.

We are very confused in our team about why the Adam optimizer does not behave the same way with the exact same codes/model/data. I also asked a colleague of mine to use the same model with his codes on my data on the DGX, validation accuracy dropped to zero for him as well. What could explain such difference while the SGD provide the same behavior? 

Hope I provided enough information, I can share more information if needed. 

Best,

**Comment by jvishnuvardhan** [comment]
@guatavita Can you please provide a standalone code with a public dataset? Issue resolution can be faster with a standalone code. Thanks!

**Comment by guatavita** [comment]
@jvishnuvardhan Can I share a private repo with you so you could replicate the issue on one of your system?

If yes, I am working on repo with a data_generator / deeplabv3+ / LiTS public database of liver segmentation.

**Comment by guatavita** [comment]
Hello @jvishnuvardhan,

Here is a google drive link of the training/validation/test set of 2D slices for the CT/masks [Google drive folder LiTs liver](https://drive.google.com/file/d/1nklG4gb9hn9JrL20d4vcqKsOB7Ps6_Ct/view?usp=sharing)

I invited you to the private repository. 

Thank you

**Comment by guatavita** [comment]
I was not able to replicate the issue with a basic binary classification model (dog/cat tutorial from Keras) and I was also not able to replicate the issue with a very simple/basic Unet model on the same data.

Following that, it seems that the issue comes from the DeepLabV3+ model. Could it be due to the Batch Normalization layers that behave differently on Windows 10/tf1.15.0 and Ubuntu/tf1.15.0+nv? 

I will push the Unet model on the private repo that I have shared.

Will try the Unet with batch norm see if it changes something.

**Comment by guatavita** [comment]
Hello,

I wanted to update this issue. We end up contacting the NVIDIA support to ask them to reproduce the behavior on a similar system (DGX server with Docker images provided by Nvidia).

The issue has been isolated, we were able to reproduce the stable results of our models using directly the tensorflow/tensorflow:1.15.2-gpu-py docker image. 

Thanks to @rmccorm4 for working on the issue.

Best,

PS: I will let the assignees close the issue.
